src/utils/lru-cache.js

```diff
@@ -16,49 +16,89 @@
 
 import {dev} from '../log';
 
-/** @typedef {{
-      payload: *,
-      access: number,
-    }} */
-export let Cacheable;
-
 /** @const {string} */
 const TAG = 'lru-cache';
 
-export class LRUCache {
-  /** @param {number} capacity */
+/**
+ * @template T
+ */
+export class LruCache {
+  /**
+   * @param {number} capacity
+   */
   constructor(capacity) {
-    /** @private {number} */
+    /** @private @const {number} */
     this.capacity_ = capacity;
 
-    /** @private {!Object<(number|string), !Cacheable>} */
+    /** @private {number} */
+    this.size_ = 0;
+
+    /**
+     * An incrementing counter to define the last access.
+     * @private {number}
+     */
+    this.access_ = 0;
+
+    /** @private {!Object<(number|string), {payload: T, access: number}>} */
     this.cache_ = Object.create(null);
   }
 
   /**
-   * @param {number|string} id
-   * @return {*} The cached payload.
+   * Returns whether key is cached.
+   *
+   * @param {number|string} key
+   * @return {boolean}
    */
-  get(id) {
-    if (this.cache_[id]) {
-      this.cache_[id].access = Date.now();
-      return this.cache_[id].payload;
+  has(key) {
+    return !!this.cache_[key];
+  }
+
+  /**
+   * @param {number|string} key
+   * @return {T} The cached payload.
+   */
+  get(key) {
+    const cacheable = this.cache_[key];
+    if (cacheable) {
+      cacheable.access = ++this.access_;
+      return cacheable.payload;
     }
     return undefined;
   }
 
   /**
-   * @param {number|string} id
-   * @param {*} payload The payload to cache.
+   * @param {number|string} key
+   * @param {T} payload The payload to cache.
+   */
+  put(key, payload) {
+    this.cache_[key] = {payload, access: this.access_};
+    this.size_++;
```

Comment:
The size doesn't increase if the key already exists, right?

Reply:
Hm, this was implicitly handled before. I'll add a check and tests.